src/filesystem.h

```diff
@@ -0,0 +1,88 @@
+/*
+ * This file is part of EasyRPG Player.
+ *
+ * EasyRPG Player is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * EasyRPG Player is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with EasyRPG Player. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef EP_FILESYSTEM_H
+#define EP_FILESYSTEM_H
+
+// Headers
+#include <cassert>
+#include <istream>
+#include <ostream>
+#include "system.h"
+
+class Filesystem {
+public:
+	class vfs_istream;
+	class vfs_ostream;
+
+	using InputStreamRaw = vfs_istream;
+	using OutputStreamRaw = vfs_ostream;
+
+	using InputStream = std::shared_ptr<InputStreamRaw>;
+	using OutputStream = std::shared_ptr<OutputStreamRaw>;
+
+	class vfs_istream : public std::istream {
+	public:
+		explicit vfs_istream(std::streambuf* sb, std::streamsize size);
+		~vfs_istream() override;
+
+		std::streamsize get_size() const;
+
+	private:
+		std::streamsize size;
+	};
```

Comment:
As we saw in https://github.com/EasyRPG/liblcf/pull/343  `seek` and `tell` result in a system call every single time they are used. For I/O which does a lot of telling, if you track the file offset yourself you get a massive speedup. For I/O which does a lot of small seeking, a call to `read()` into an empty buffer can be faster than a small seek.

If we wanted to fix this at the stream level, it would have to be done by implementing our own custom replacement for `filebuf`. Something to consider for future PR's. Maybe it could speed up the audio system?

Once all this work is done, an optimized `filebuf` could be worth exploring.

Reply:
I think this will be useful to implement. Other streams like compressed archive streams that must be baked by memory for performance will have a fast tell so "Real" Filestreams should have the same improvements
Also even zip reading has the fstream as underlying stream so everything will profit from it.

Reply:
For the file only implementation, one might be able to get away with simply creating a new streambuf which wraps `filebuf` and intercepts it's seek and tell calls. This would need to be serparately unit tested and benchmarked carefully. I would definatly save any attempt for later PR.

For the zip file version, the details are notfully clear to me. But the overall design you'd want separate stream objects for each use like the file case. Each of those streams can reference a shared backend with 1 file handle. The buffer could be an LRU or ARC cache shared by all streams, or each stream could have it's own buffer. For this use case having a `shared_ptr` hidden inside the stream class could make a lot of sense. The zipfile resources go away exactly when all streams are closed.

Or would it just be easier to open same zipfile multiple times as read-only? One could cache and index to the file somewhere so new streams can jump directly to the part of the file they need to read without scanning.